schema-engine/core/src/state.rs

```diff
@@ -523,4 +524,34 @@ impl GenericApi for EngineState {
         }))
         .await
     }
+
+    async fn dispose(&mut self) -> CoreResult<()> {
+        self.connectors
+            .lock()
+            .await
+            .drain()
+            .map(|(_, snd)| async move {
+                let (tx, mut rx) = mpsc::channel(1);
+
+                let send_result = snd
+                    .send({
+                        let tx = tx.clone();
+                        Box::new(move |conn| {
+                            Box::pin(async move {
+                                _ = tx.send(conn.dispose().await).await;
+                            })
+                        })
+                    })
+                    .await;
+
+                if send_result.is_err() {
+                    _ = tx.send(Ok(())).await;
+                }
+
+                rx.recv().await.unwrap_or(Ok(()))
```

Comment:
This statement stood out to me. How is this safe?

Reply:
Result can be `None` if the sender was dropped (which isn't even possible at all even if the closure was dropped without running because we keep our copy of `tx`). Even if it were dropped, we'd want to collect the result for this connector as whatever so that we don't wait for it infinitely. It's similar to the `if` statement above, where we collect `Ok` if somehow the mailbox for the connector was dropped before we even told it to `dispose`.

None if this should be possible in practice since `dispose` requires exclusive access to `State` via a mutable reference, so we could as well invent some error types here or even panic, but I decided to keep it simple and lenient since it's shutdown logic anyway.